### PR TITLE
Permission handling cleanup

### PR DESCRIPTION
**Pull request by adrianschroeter**
first part of fixes, some more to come, but that need a bit more time

### CODE REVIEW

**Comment by bgeuken** [comment]
No tests?;-)

**Comment by adrianschroeter** [comment]
the second part will come with test case but they show only the breakage atm :/

A testcase which could be written here is the project creation via webui when configuration.allow_user_to_create_home_project = false is set.....

**Comment by bgeuken** [comment]
Feel free to take this one: https://github.com/bgeuken/open-build-service/tree/fix_issue_1291
Though there might be more things to test.

**Comment by adrianschroeter** [comment]
This is only testing if subprojects can be created, we need to test that "tom" is allowed to create "home:tom" in one case and not in the other one.

The subproject creation needs always to work when "home:tom" exists.

**Comment by adrianschroeter** [comment]
but you have convienced me to adapt it :)

**Comment by bgeuken** [comment]
@adrianschroeter I know. I was working on this issue to and started with subprojects. That's why I only had a test for that case so far.

File: src/api/app/controllers/source_controller.rb [hunk 1]
```diff
@@ -544,8 +544,8 @@ def update_project_meta
       else
         project = Project.new(name: project_name)
         project.update_from_xml!(request_data)
-        # failure is ok
-        project.add_user(User.current.login, 'maintainer')
+        # FIXME3.0: don't modify send data
+        project.relationships.build(user: User.current, role: Role.find_by_title!('maintainer'))
```
Comment (hennevogel) [inline src/api/app/controllers/source_controller.rb]: So if I update the meta file I automatically become maintainer?

File: src/api/test/functional/webui/create_test.rb [hunk 2]
```diff
@@ -17,16 +17,29 @@ def test_create_package
...
     fill_in 'project_name', :with => 'coolstuff'
     click_button 'Create Project'
     flash_message.must_equal "Project 'home:tom:coolstuff' was created successfully"
+    assert Project.where(name: "home:tom:coolstuff").exists?
 
     assert current_url.end_with?(project_show_path(project: 'home:tom:coolstuff')), "#{current_url} not ending with coolstuff"
     find('#project_title').text.must_equal 'home:tom:coolstuff'
     first('#packages li').text.must_equal 'Packages (0)'
+
+    Configuration.stubs(:allow_user_to_create_home_project).returns(false)
```
Comment (hennevogel) [inline src/api/test/functional/webui/create_test.rb]: Please add a second test case for this...

File: src/api/test/functional/webui/project_controller_test.rb [hunk 3]
```diff
@@ -505,6 +505,18 @@ def test_rebuild_time_on_apache
...
+    Configuration.stubs(:allow_user_to_create_home_project).returns(false)
+    visit new_project_path
+    fill_in 'project_name', with: 'home:user1'
+    click_button('Create Project')
+    assert_equal count, Project.count
+    flash_message.must_equal "Sorry you're not allowed to create this Project"
+    flash_message_type.must_equal :alert
+
+    # allow creation and retry
+    Configuration.stubs(:allow_user_to_create_home_project).returns(true)
```
Comment (hennevogel) [inline src/api/test/functional/webui/project_controller_test.rb]: Please add a second test case for this

File: src/api/app/controllers/source_controller.rb (hunk 1 above)
Comment (adrianschroeter) [inline src/api/app/controllers/source_controller.rb]: On Freitag, 23. Oktober 2015, 06:13:11 CEST wrote Henne Vogelsang:
> > @@ -544,8 +544,8 @@ def update_project_meta
> >        else
> >          project = Project.new(name: project_name)
> >          project.update_from_xml!(request_data)
> > -        # failure is ok
> > -        project.add_user(User.current.login, 'maintainer')
> > +        # FIXME3.0: don't modify send data
> > +        project.relationships.build(user: User.current, role: Role.find_by_title!('maintainer'))
> 
> So if I update the meta file I automatically become maintainer?

no, this is the case for creating a new project only.

**Comment by adrianschroeter** [comment]
Björn, JFYI, I dropped your commit again since it broke the create_subproject test, I did not debug it though

**Comment by bgeuken** [comment]
No problem. Once this got merged I am checking the test again and create a new PR. We should cover the subproject workflow.

**Pull request merged**